runtime/stdlib.js

```diff
@@ -352,85 +352,129 @@ function caml_compare_val (a, b, total) {
+          // TODO: We should rewrite the implementation and follow
+          // what we do for other custom blocks: zarith, bigint, bigarray, ...
+          var x = caml_int64_compare(a, b);
+          if (x != 0) return x;
+          break;
+        case 1247: // Function
+          caml_invalid_argument("compare: functional value");
+          break;
+        case 1255: // Custom
+          var comp = (caml_custom_ops[a.caml_custom] && caml_custom_ops[a.caml_custom].compare);
+          if(comp != (caml_custom_ops[b.caml_custom] && caml_custom_ops[b.caml_custom].compare)){
+            return (a.caml_custom<b.caml_custom)?-1:1;
           }
-        } else
-          // Should not happen
-          return 1;
-      } else if (a instanceof Array && a[0] === (a[0]|0)) {
-        var ta = a[0];
-        // ignore double_array_tag
-        if (ta === 254) ta=0;
-        // Forward object
-        if (ta === 250) {
-          a = a[1];
-          continue;
-        } else if (b instanceof Array && b[0] === (b[0]|0)) {
-          var tb = b[0];
-          // ignore double_array_tag
-          if (tb === 254) tb=0;
-          // Forward object
-          if (tb === 250) {
-            b = b[1];
-            continue;
-          } else if (ta != tb) {
-            return (ta < tb)?-1:1;
-          } else {
-            switch (ta) {
-            case 248: {
-              // Object
-              var x = caml_int_compare(a[2], b[2]);
-              if (x != 0) return x;
-              break;
-            }
-            case 251: {
-              caml_invalid_argument("equal: abstract value");
-            }
-            case 255: {
-              // Int64
-              var x = caml_int64_compare(a, b);
-              if (x != 0) return x;
-              break;
-            }
-            default:
-              if (a.length != b.length) return (a.length < b.length)?-1:1;
-              if (a.length > 1) stack.push(a, b, 1);
-            }
-          }
-        } else
-          return 1;
-      } else if (b instanceof MlBytes ||
-                 (b instanceof Array && b[0] === (b[0]|0))) {
-        return -1;
-      } else if (typeof a != "number") {
-        if(a && a.compare) {
-          var cmp = a.compare(b,total);
-          if (cmp != 0) return cmp;
-        }
-        else if(a && a.caml_custom) {
-          if(caml_custom_ops[a.caml_custom].compare) {
-            var cmp = caml_custom_ops[a.caml_custom].compare(a,b,total)
-            if (cmp != 0) return cmp;
+          if(!comp)
+            caml_invalid_argument("compare: abstract value");
+          var x = comp(a,b,total);
+          if (x != 0) return x;
+          break;
+        case 1256: // compare function
+          var x = a.compare(b,total);
+          if(x != x | 0)
+            caml_invalid_argument("compare: abstract value");
+          x = x | 0;
+          if (x != 0) return x;
+          break;
+        case 1000: // Number
```

Comment:
What happens when integers are compared against floats?

Reply:
Well, it should behave like comparison on floats